### ISSUE TITLE
Reset Credentials Flow logs in without OTP

### ISSUE THREAD

**Issue by abstractj** [keycloak/keycloak#12759]
## Original report
A user has password and totp as credentials. The reset credentials flow is configured such that reset totp is disabled, the totp should not be modified

2 Scenarios:

1. 

- User operns browser, gets to login screen, selects reset password and enters username
- User opens email, copies link to same browser instance as used for triggering the reset
- and enters new password - and is logged in !!, without having to enter otp

2.

- User operns browser, gets to login screen, selects reset password and enters username
- User opens email, copies link to a new browser instance
- and enters new password - and then everything is ok, password is reset, user has to login

Expected behavior:
Even in scenario 1, i.e same browser instance for reset password trigger and opening link and entering new password, the user has to start a new login before he is actually logged in, --- or ---, that the the user has to enter the correct otp

## Steps to reproduce

- Create user with password and totp
- Configure Reset Credential flow such that reset totp is disabled
- Open account url for that user and proceed as described in scenario 1

## Mitigation

A possible workaround to mitigate this, while we don't have the fix for it, is to include the OTP form as part of the reset credentials form, if you would like. Below you can find the screenshot with the suggested configuration.

![mitigation](https://user-images.githubusercontent.com/21758/175646562-b0017117-115c-45b4-b57e-669cd3c356c9.png)

## Version
- <= 18.0.1

## References
- https://issues.redhat.com/browse/KEYCLOAK-18885
- [Reset Credentials Improvements](https://issues.redhat.com/browse/KEYCLOAK-13134)

## Reporter
- Stefan Albrecht

**Comment by abstractj** [comment]
After discussing it in the Keycloak security, we agreed that this is a security hardening. Quoting @mposolda: "This is security hardening IMO and it should be addressed in the wider context during the work on general improvements for "Reset Credentials" functionality.

There was discussion about this on keycloak-security with the topic: `Default "Reset Credentials Flow" is insecure`"

**Comment by michalpomykala** [comment]
Hello, any update on that? We struggling the same problems on production servers. Could you be so kind and give some feedback what is the status ?

**Comment by michalpomykala** [comment]
@abstractj do you have any solution for that? How did you manage this problem?

**Comment by mposolda** [comment]
Changing priority to "normal" and added it to the epic https://github.com/keycloak/keycloak/issues/27732 for authentication policies.

This does not have any straightforward and flexible solution for now and should be addressed as part of wider effort of improving authentication with authentication policies.

@michalpomykala In the meantime, you can use the workaround added in the "Mitigation" section of this issue. You can add "OTP Form" as REQUIRED authenticator and disable the `Copy of reset ...` subflow. In this case, user will be required to fill OTP before login instead of being asked to update his OTP.

**Comment by NikolaNemes** [comment]
Hi everyone,

Just to double-check since I ran into this issue recently:
- The mitigation specified in this is no longer valid since Keycloak now prohibits editing of default/built-in flows
- - https://github.com/keycloak/keycloak/issues/15536
- Another option would be to duplicate the flow, alter the copy and use the copy
- - However it is also no longer possible to override the reset credentials flow?
- - In the Admin UI I can only see options to override the browser flow and direct grant flow, whereas in older versions it was possible to override the reset credentials flow as well

Perhaps I am also wrong with my assumptions, but what would be the recommended way to handle this with the newer versions of keycloak?

**Comment by NikolaNemes** [comment]
Seems it's possible to override the UI was changed and I wasn't able to figure it out at all.

For anyone coming up on this issue:
- Click on the kebab menu next to the listed flows and select 'Bind flow'
- Or while editing a single flow use the action button (This action only appears for non built-in flows)

**Comment by codeindustriesnl** [comment]
It seems the same is true for the flow that triggers the email validation if TOTP is only on the browser flow. This browser flow is never triggered when validating the email address.

**Comment by andy-clapson** [comment]
I agree that this should be the default, users should have to edit Flows to make things LESS secure, not the other way around.

I think the bigger question to solve for features like this is how to allow a migration path for default flows. If a KC instance has been running for years, the flows don't move at upgrade time, even if defaults are updated for the better.
